### PR TITLE
GitHub like repo home page

### PR DESCRIPTION
**Pull request by yp05327**
In the repo home page of GitHub, user can quickly get the latest release, contributors, and other useful information.

And the main point is that we can:
- add contributors page which will be provided in #27882
- add GitHub's activity page (the activity page in Gitea is called Insights in GitHub)
- add quick downloading button for the latest release
- move the language status bar to the sidebar, then user can easily know what is that green( or other colors) bar

In desktop view:
image
In mobile view: (will not show readme button and `Description` title)
image
image

### CODE REVIEW

**Comment by lng2020** [comment]
Will the language status bar move to the right side too? I always think the language status bar goes unnoticed for someone new to Gitea.

**Comment by JakobDev** [comment]
We already have all this Information visible on the Site.

**Comment by JakobDev** [comment]
I'm personal more in favour of the current concept. We have Description and Topics of the Repo on Top, which makes IMHO way more sense than on the Left.

**Comment by silverwind** [comment]
Many were against this design but I guess it had to happen at some point 😆.

**Comment by delvh** [comment]
One thing I've noticed on your screenshot:
I think we should display the `release` above the stats.
Most often, people want to see the release.
It is rare that you want to see how many stars a project has.

**Comment by yp05327** [comment]
I think maybe we don't need stars/forks/watching in the sidebar as they are already displayed above.

And the main point is that I want to:
- add contributors page
- add GitHub's activity page (the activity page in Gitea is called `Insights` in GitHub) 
- add quick downloading the latest release
- move the language status bar to the sidebar, then user can easily know what is that green( or other colors) bar

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -1053,6 +1053,13 @@ form.reach_limit_of_creation_n = The owner has already reached the limit of %d r
 form.name_reserved = The repository name "%s" is reserved.
 form.name_pattern_not_allowed = The pattern "%s" is not allowed in a repository name.
 
+sidebar.readme = Readme
+sidebar.stars = stars
+sidebar.forks = forks
+sidebar.watching = watching
+sidebar.contributors = Contributors
+latest_release = Latest
```
Comment (silverwind) [inline options/locale/locale_en-US.ini]: Move all these translation to top-level imho to make them re-usable, e.g. translation key `readme`, `stars` etc.

File: templates/repo/sidebar/repo_info.tmpl [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+					<div>
+						<button class="ui basic button" id="cancel_topic_edit">{{ctx.Locale.Tr "cancel"}}</button>
+						<button class="ui primary button" id="save_topic" data-link="{{.RepoLink}}/topics">{{ctx.Locale.Tr "save"}}</button>
+					</div>
+				</div>
+				{{end}}
+			</div>
+		</div>
+	</div>
+{{end}}
```
Comment (silverwind) [inline templates/repo/sidebar/repo_info.tmpl]:
```suggestion
{{end}}

```

**Comment by silverwind** [comment]
Inside the `...` menu is a "Cite this repository", it should be moved to sidebar as well, to the "stars" section.

**Comment by silverwind** [comment]
Yeah we can omit duplicate info like stars/forks.

**Comment by yp05327** [comment]
> Inside the `...` menu is a "Cite this repository", it should be moved to sidebar as well, to the "stars" section. You need a repo like [this one](https://github.com/silverwind/symlink-test) with a CITATION file.

The div of this  `...` menu has a display problem on mobile view.
image

File: options/locale/locale_en-US.ini (hunk 1 above)
Comment (yp05327) [reply options/locale/locale_en-US.ini]: Done.

**Comment by yp05327** [comment]
> Cite this repository

It is new for me, and it seems that we have an issue for this feature.
image

https://gitea.com/yp05327/testrepo

**Comment by silverwind** [comment]
Definitely a bug. Does not happen on my test repo: https://try.gitea.io/silverwind/symlink-test

**Comment by yp05327** [comment]
> Definitely a bug. Does not happen on my [test repo](https://try.gitea.io/silverwind/symlink-test), even though the file content is the same.

I can reproduce this on my local test server which is using the latest version.

**Comment by yp05327** [comment]
This error comes from line 17.
image
Then I print `citationFileContent`, it seems that `date-released` is incorrect.
image

**Comment by yp05327** [comment]
I know, we only read the first 1024 bytes of this file. So the content is cut off.
image

**Comment by yp05327** [comment]
Fix #27958

**Comment by yp05327** [comment]
We can add a link to access the contributors page which will be provided in #27882.

**Comment by yp05327** [comment]
Do we have this? IIRC, we don't have this.
image

**Comment by silverwind** [comment]
> Do we have this? IIRC, we don't have this. image

No, I don't think we have this yet.

**Comment by JakobDev** [comment]
The big problem I still see here, is that this design takes  up a lot vertical space. the old design takes only as much horizontal space as needed, but the new design will take vertical space from the full page.

**Comment by triggered96** [comment]
> Fix #27958

截圖 2023-11-28 晚上8 41 35

**Comment by yp05327** [comment]
> > Fix #27958
> 
> 截圖 2023-11-28 晚上8 41 35
> Search box bug

Thanks for your test!
Can you provide some details about this issue？
I have no idea about how to reproduce it. 🤔

**Comment by denyskon** [comment]
@yp05327 Do you have code indexer enabled? Then a search bar appears below the repo description normally

**Comment by triggered96** [comment]
> > > Fix #27958
> > 
> > 
> > 截圖 2023-11-28 晚上8 41 35
> > Search box bug
> 
> Thanks for your test! Can you provide some details about this issue？ I have no idea about how to reproduce it. 🤔

```
# vi config/app.ini 

[indexer]
;; repo indexer by default disabled, since it uses a lot of disk space
REPO_INDEXER_ENABLED = true
```

**Comment by triggered96** [comment]
Another bug is `<button id="manage_topic" class="btn interact-fg gt-font-12">Manage Topics</button>`. There is no response when pressing the button

**Comment by yp05327** [comment]
There are two `manage topic` button. So only the first one have response which is used for mobile view.

**Comment by yp05327** [comment]
I think these two issues are all fixed.

**Comment by triggered96** [comment]
> I think these two issues are all fixed.

Just tested it, the information effect is very good.👌

**Comment by denyskon** [comment]
I don't think placing the search bar in the navbar is a good idea. What is wrong with the old location?

**Comment by yp05327** [comment]
> I don't think placing the search bar in the navbar is a good idea. What is wrong with the old location?

After this change, we can search in each page but not only in the home page.

**Comment by yp05327** [comment]
> > I think these two issues are all fixed.
> 
> 截圖 2023-11-29 晚上9 30 50

This place has many problems. e.g. https://github.com/go-gitea/gitea/pull/27931#issuecomment-1797068112
Have no idea about how to fix it. And they are not completely related to this PR, you can also find these problems in the latest version. I think we can fix them in another PR.

**Comment by denyskon** [comment]
@yp05327 I think we shouldn't move one specific search bar for one specific page into nav bar. We could do it with every search bar inside Gitea (use one standard search bar like GitHub or GitLab does), however that would be a bigger refactoring.

**Comment by yp05327** [comment]
> @yp05327 I think we shouldn't move one specific search bar for one specific page into nav bar. We could do it with every search bar inside Gitea (use one standard search bar like GitHub or GitLab does), however that would be a bigger refactoring.

I also talked about this to @CaiCandong yesterday. So just move it above the `Description` ? And then improve it later.

**Comment by oilrich25** [comment]
> > @yp05327 I think we shouldn't move one specific search bar for one specific page into nav bar. We could do it with every search bar inside Gitea (use one standard search bar like GitHub or GitLab does), however that would be a bigger refactoring.
> 
> I also talked about this to @CaiCandong yesterday. So just move it above the `Description` ? And then improve it later.


Certainly! Here's the translated English version of the commit response:

Thank you for your contribution! We appreciate your efforts on the GitHub like repo home page.

I've noticed that your pull request doesn't seem to be compatible with the new version of Gitea 1.21.1. It appears that the author's GitHub like repo home page optimizations are primarily geared towards the older version, and some adjustments may be needed for certain changes in the new version.

I suggest checking the documentation for the latest version of Gitea to see if there are any updates regarding the GitHub like repo home page. This will help ensure that your pull request aligns with the latest version. If you need further assistance or guidance, please feel free to let me know.

**Comment by algernon** [comment]
I'm loving these changes - came up with something similar on my own, before being pointed to this PR, and this is - in many ways - much better. One thing I did that isn't done here, and might be worth considering, is getting rid of the "N commits, X branches, Z tags, S size" bar at the top of the repo: move the commits to the repo view, branches & tags inbetween the branch selector and the clone widget, and the size to the sidebar, for example.

Something like this:

Screenshot 2023-12-23 at 10-52-56 riemann-c-client

**Comment by yp05327** [comment]
As #27882 merged, we can restart to finish this PR.

**Comment by yp05327** [comment]
There's a problem that the current approach is that contributor data will be generated dynamically, when user access contributor page, and save it into cache. So no data will be stored in db.
And the sort is executed by js in the browser, but we can not do this in the repo home page. 🤔

**Comment by denyskon** [comment]
@sahinakkaya Could you help here? :)

**Comment by sahinakkaya** [comment]
I was exposing an endpoint to generate data but we decided to remove it after discussing it here: https://github.com/go-gitea/gitea/pull/27882#discussion_r1464233021. Relevant commit is [here](https://github.com/go-gitea/gitea/commit/3740877851e9cddb185289af7498d748e9ec7d11). Maybe we can open another PR to include it again?

**Comment by sahinakkaya** [comment]
On a second thought, api shouldn't be necessary. Calling this function from your code should get you the data. No? 
https://github.com/go-gitea/gitea/blob/7132a0ba75d6fe734d9f950f217a5ceb81375328/services/repository/contributors_graph.go#L82

In `/routers/web/repo/contributors.go`, I am calling it like this:

https://github.com/go-gitea/gitea/blob/7132a0ba75d6fe734d9f950f217a5ceb81375328/routers/web/repo/contributors.go#L33-L43

**Comment by yp05327** [comment]
> On a second thought, api shouldn't be necessary. Calling this function from your code should get you the data. No?
> 
> https://github.com/go-gitea/gitea/blob/7132a0ba75d6fe734d9f950f217a5ceb81375328/services/repository/contributors_graph.go#L82
> 
> In `/routers/web/repo/contributors.go`, I am calling it like this:
> 
> https://github.com/go-gitea/gitea/blob/7132a0ba75d6fe734d9f950f217a5ceb81375328/routers/web/repo/contributors.go#L33-L43

Yes, but the data are not sorted. By my test, the order is random.
And there's a key called `total` in `ContributorStats` which seems should be another variable to return, or we need to add a if condition and the real count of contributors of `ContributorStats` should minus 1.
You can check a draft changes in https://github.com/go-gitea/gitea/commit/4eec6646fd9cd3ef6f7c9fd00f2d27fcd6304034#diff-43a7f4747c7ba8bff888c9be11affaafd595fd55d27f3333840eb19df9fad393R903

I'm not sure which is the best way to store these data. In DB, then we can use SQL to get sorted data instead of doing this by Gitea, but how to update? Or in cache, then we need cost CPU time and memory to do the sort.

**Comment by yp05327** [comment]
Repo home page only displays top contributors not all of them, in GitHub, it is 14.
Maybe we should create a cron job to sync contributors data into DB every day or week. Then we can easily get them from DB. 🤔

**Comment by sahinakkaya** [comment]
Yeah, creating a cron job is good idea :+1:

**Comment by silverwind** [comment]
FWIW, GitLab seems to also recently have switch to this layout so I'm warming up to the idea 😆 

image

**Comment by yp05327** [comment]
@silverwind 
Done

**Comment by yekanchi** [comment]
I think this red part can be removed.
image

**Comment by denyskon** [comment]
Please don't, at least commits and branches page should be accessible from the top I think.

**Comment by silverwind** [comment]
I guess we'd need squeeze everything into first line which will be a lot of work, likely something for later.

**Comment by yp05327** [comment]
I prefer move repo size to right, and hold rest of them.
It is repo's information, we can move it to the right where is used to display the information.
But rest of them are quick links. they should be there, or maybe they can be in one line or other similar designs.

**Comment by yp05327** [comment]
Too many conflicts, maybe broken now.

**Comment by yp05327** [comment]
`gw-font-semibold` has been converted in to `tw-font-semibold`
but I can not find the css definition.
image

**Comment by yp05327** [comment]
I think I need to rewrite it as there are too many things changed. 😕

**Comment by silverwind** [comment]
> `gw-font-semibold` has been converted in to `tw-font-semibold` but I can not find the css definition.

`tw-font-semibold` is correct, You can use [original Tailwind classes](https://tailwindcss.com/docs/font-weight). We map them to the CSS vars which was done in https://github.com/go-gitea/gitea/pull/30027.

**Comment by yp05327** [comment]
Contributor stats only uses cache and it will always re-generate when the server restarts.
And it returns map so the order will be random and also include unrelated summary data in it.
There are too many problems to implement contributors in this PR, so I decide to remove it unless it can be easily be reused in other places.

**Comment by silverwind** [comment]
> Contributor stats only uses cache and it will always re-generate when the server restarts. And it returns map so the order will be random and also include unrelated summary data in it. There are too many problems to implement contributors in this PR, so I decide to remove it unless it can be easily be reused in other places.

Yeah, keep the changeset minimal so it only does the layout change.

**Comment by denyskon** [comment]
@yp05327 What's the state of this PR?

**Comment by yp05327** [comment]
The frontend has huge changes in recent, I think this should be rewritten.

**Comment by yp05327** [comment]
#15626 this can also be included.

**Comment by yp05327** [comment]
As #24872 merged, the next step is replacing this PR.

**Comment by yp05327** [comment]
Replaced by #32213

**Pull request closed without merging**